### ISSUE TITLE
Universal Wrapper Device fails to recognize device type of devices with custom channels

### ISSUE THREAD

**Issue by mulder999** [jens-maus/cuxd#6]
Referencing channel 3 of HM-LC-Sw4-DR does not recognize channel type. While it does work for the other 3 channels:

	CUX9001004:	CCU(SWITCH,'KEQ084XXXX:2.STATE') trigger(2:1.000)
	CUX9001005:	CCU(SWITCH,'KEQ084XXXX:1.STATE') trigger(2:1.000)
	CUX9001007:	CCU(SWITCH,'KEQ084XXXX:4.STATE') trigger(2:1.000)
	CUX9001008:	CCU((null),'KEQ084XXXX:3.STATE') trigger(2:1.000)

Affects Version 1.10b1 on RaspberryMatic 2.29.18.20170731.

**Comment by uwe111** [comment]
This rather looks like a ReGaHss problem with your device.
Try the following HM-script to verify that.
`
string c;
foreach (c,root.Channels().EnumUsedIDs()){
  var c1 = dom.GetObject(c);
  if (c1.Address().Find('KEQ084XXXX') != -1) {
    var c2 = dom.GetObject(c1.Interface());
    WriteLine(c2.Name() # '|' # c1.Address() # '|' # c1.HssType());
  }
}
`

**Comment by mulder999** [comment]
Thanks for looking, result is:
BidCos-RF|KEQ084XXXX:1|SWITCH
BidCos-RF|KEQ084XXXX:2|SWITCH
BidCos-RF|KEQ084XXXX:3|SWITCH
BidCos-RF|KEQ084XXXX:4|SWITCH

**Comment by mulder999** [comment]
I noticed I have attached additional custom datapoints to this channel. Here is an extract of XMLAPI statelist.cgi.

<channel name="Photo:Heat:Valve" ise_id="20409" visible="true" operate="true">
<datapoint name="BidCos-RF.KEQ084XXXX:3.STATE" type="STATE" ise_id="20413" value="false" valuetype="2" valueunit="" timestamp="1502662384" operations="7"/>
<datapoint name="Var:Photo:Heat:Valve:Accumulator" type="" ise_id="20008" value="0.000000" valuetype="4" valueunit="/20" timestamp="1502456692" operations="7"/>
<datapoint name="Var:Photo:Heat:Valve:Order" type="" ise_id="20004" value="1.000000" valuetype="4" valueunit="/20" timestamp="1502456692" operations="7"/>
</channel>

**Comment by mulder999** [comment]
Dropping and recreating the custom data points after CUxD has made its checks seems a working workaround (until next reboot). Hence suggesting to reopen this issue to improve the datapoint matching algorithm in CUxD.

**Comment by uwe111** [comment]
I don't see a problem with datapoint matching here.
What are the CUxD Syslog messages after you reboot your CCU?
There should be lines like this in the log:

> ... setup CUX9001004: ...
> ...
> ...

Maybe the channels are not fully available directly after the CCU reboot?
The wrapper device should work anyway, or doesn't it?

**Comment by mulder999** [comment]
Thanks for pointing into direction, the error in Syslog reads:

```
Aug 15 20:24:53 homematic-raspi local0.err ReGaHss: Error: IseESP::ScriptRuntimeError: string c; string ret; foreach (c,root.Channels().EnumUsedIDs()){ var c1 = dom.GetObject(c); if (c1.Address().Find('KEQ084XXXX:3') != -1) { var c2 = dom.GetObject(c1.Interface()); ret = c2.Name() # '|' # c1.Address() # '|' # c1.HssType() # '|'; string dp; foreach (dp,c1.DPs()) { ret = ret # dom.GetObject(dp).HssType() # '|'; } Write(ret); quit; } }  [iseESPexec.cpp:12344]
```

Indeed dom.GetObject(dp).HssType() returns nothing for custom datapoint (that is variable assigned to a channel), hence the concatenation fails. A bit unsure what part is generating that script...

**Comment by mulder999** [comment]
I can confirm, this affects only the logs and user interface (the 2 read only fields used for device configuration feedback).

**Comment by uwe111** [comment]
CUxD generates this script, so maybe it has to check dom.GetObject(dp).HssType() for existence first?
Since I don't have custom datapoints, can you please check if that would work?
Or maybe use another method to return the datapoints?
The HM-script is defined in cuxd.inc, section ##15
Please let me know the results.

**Comment by mulder999** [comment]
Thanks for pointer.
We can check that the DP object has the TypeName "HSSDP" (btw custom DP have the type VARDP), see branch [fix-VARDP](https://github.com/mulder999/cuxd/tree/fix-VARDP). For the moment, I am still not clear of the actual meaning of '|' char in section ##28 start of lines 503, 504 and 507, hence waiting until the fix is complete before creating the pull request...

**Comment by uwe111** [comment]
The '|' lines are comment lines. They are ignored.
section ##28 is used for a similar (older) function with a different return value in CUxD device (05) FS20 Relais.

**Comment by mulder999** [comment]
Thanks, was thinking it was a complex way of doing conditional statement's injections ;-)
Issue shall now be fixed by pull request #7 .

Regards

**Comment by jens-maus** [comment]
@mulder999 Thanks for this nice fix for recent CCU firmwares (2.29.18+)

@uwe111 please take over this file into your own build environment for the CCU1+CCU2 builds of future CUxD versions.
